src/table.js

```diff
@@ -684,10 +762,102 @@ export function __table(source, operations) {
       }))
     );
   }
-  if (primitive) source = source.map((d) => d.value);
   if (source !== input) {
     if (schema) source.schema = schema;
     if (columns) source.columns = columns;
   }
   return source;
 }
+
+function coerceRow(object, types, schema) {
+  const coerced = {};
+  for (const col of schema) {
+    const type = types.get(col.name);
+    const value = object[col.name];
+    coerced[col.name] = type === "raw" ? value : coerceToType(value, type);
+  }
+  return coerced;
+}
+
+function initKey() {
+  return {
+    other: 0,
+    boolean: 0,
+    integer: 0,
+    number: 0,
+    date: 0,
+    string: 0,
+    array: 0,
+    object: 0,
+    bigint: 0,
+    buffer: 0
+  };
+}
+
+// We need to show *all* keys present in the array of Objects
+function getAllKeys(rows) {
+  const keys = new Set();
+  for (const row of rows) {
+    // avoid crash if row is null or undefined
+    if (row) {
+      // only enumerable properties
+      for (const key in row) {
+        // only own properties
+        if (Object.prototype.hasOwnProperty.call(row, key)) {
+          // unique properties, in the order they appear
+          keys.add(key);
+        }
+      }
+    }
+  }
+  return Array.from(keys);
+}
+
+export function inferSchema(source, columns = getAllKeys(source)) {
+  const schema = [];
+  const sampleSize = 100;
+  let sample = source.slice(0, sampleSize);
+  const typeCounts = {};
+  for (const d of sample) {
+    for (const col of columns) {
+      if (!typeCounts[col]) typeCounts[col] = initKey();
+      const type = typeof d[col];
+      const value = type === "string" ? d[col].trim() : d[col];
+      if (type !== "string") {
+        if (Array.isArray(value)) ++typeCounts[col].array;
+        else if (value instanceof Date) ++typeCounts[col].date;
+        else if (value instanceof ArrayBuffer) ++typeCounts[col].buffer;
+        else if (type === "number") {
+          ++typeCounts[col].number;
+          if (Number.isInteger(+value)) ++typeCounts[col].integer;
```

Comment:
```suggestion
          if (Number.isInteger(value)) ++typeCounts[col].integer;
```

(Here we already know that `value` is a number so no coercion is necessary.)